add hardsigmoid FP operator to PyTorch

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34545 add hardsigmoid FP operator to PyTorch**

Summary:

This is for common operator coverage, since this is widely used.  A future PR
will add the quantized version.

Some initial questions for reviewers, since it's my first FP operator
diff:
* do we need a backwards.out method for this?
* do we need CUDA? If yes, should it be this PR or is it ok to split

Test Plan:

```
// test
python test/test_torch.py TestTorchDeviceTypeCPU.test_hardsigmoid_cpu_float32

// benchmark
python -m pt.hardsigmoid_test
...
Forward Execution Time (us) : 40.315

Forward Execution Time (us) : 42.603
(full output: https://gist.github.com/vkuzo/4ab3067a3466a8155c6b07c57b346663)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20371692](https://our.internmc.facebook.com/intern/diff/D20371692)

vkuzo committed:
add hardsigmoid FP operator to PyTorch

Summary:

This is for common operator coverage, since this is widely used.  A future PR
will add the quantized version.

Some initial questions for reviewers, since it's my first FP operator
diff:
* do we need a backwards.out method for this?
* do we need CUDA? If yes, should it be this PR or is it ok to split

Test Plan:

```
// test
python test/test_torch.py TestTorchDeviceTypeCPU.test_hardsigmoid_cpu_float32

// benchmark
python -m pt.hardsigmoid_test
...
Forward Execution Time (us) : 40.315

Forward Execution Time (us) : 42.603
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "add hardsigmoid FP operator to PyTorch"


Summary:

This is for common operator coverage, since this is widely used.  A future PR
will add the quantized version.

Some initial questions for reviewers, since it's my first FP operator
diff:
* do we need a backwards.out method for this?
* do we need CUDA? If yes, should it be this PR or is it ok to split

Test Plan:

```
// test
python test/test_torch.py TestTorchDeviceTypeCPU.test_hardsigmoid_cpu_float32

// benchmark
python -m pt.hardsigmoid_test
...
Forward Execution Time (us) : 40.315

Forward Execution Time (us) : 42.603
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20371692](https://our.internmc.facebook.com/intern/diff/D20371692)

[ghstack-poisoned]

jamesr66a commented on aten/src/ATen/native/cpu/Activation.cpp:
```diff
@@ -314,6 +314,65 @@ void GeluBackwardKernelImpl(TensorIterator& it) {
+    const scalar_t three(3.0f);
+    const scalar_t six(6.0f);
+    using Vec = vec256::Vec256<scalar_t>;
+    const Vec kZeroVec(0);
```
Can we add explicit casts to `scalar_t` on the arguments to these constructors?

jamesr66a commented on aten/src/ATen/native/cpu/Activation.cpp:
```diff
@@ -314,6 +314,65 @@ void GeluBackwardKernelImpl(TensorIterator& it) {
+            : zero;
+          return (
+            val_plus_three_gt_zero < six ? val_plus_three_gt_zero : six
+          ) / six;
```
Can't this be implemented branch-free like `std::min(std::max(x + three, zero), size) / six`

jamesr66a commented on aten/src/ATen/native/cpu/Activation.cpp:
```diff
@@ -314,6 +314,65 @@ void GeluBackwardKernelImpl(TensorIterator& it) {
+            valPlusThreeGTZero,
+            kSixVec,
+            valPlusThreeGTZero > kSixVec
+          ) / kSixVec;
```
Same here, branch free but replace

std::min/max -> vec256::minimum/maximum
three, zero, six -> kThree/Zero/SixVec

jamesr66a commented on torch/nn/functional.py:
```diff
@@ -1569,6 +1569,25 @@ def sigmoid(input):
+
+    See :class:`~torch.nn.Hardsigmoid` for more details.
+    """
+    # TODO: question for reviewers, is this JIT portion correct?
```
Can you elaborate on why this special branch is needed?

vkuzo replied:
would love to know as well, all the other functions in this file have it so I made an educated guess that it's needed.  @raghuramank100 , any ideas on who might know?

jamesr66a replied:
Oh when i look into it it's to not do the whole `__torch_function__` business if we're doing TorchScript: https://github.com/pytorch/pytorch/commit/fa80299bdf111ec3bc1ed1c4b3492bfe449817c7

cc @suo this is probably something that should be on our radar as something that TS doesn't support

jamesr66a commented on test/test_torch.py:
```diff
@@ -14132,6 +14132,24 @@ def test_sigmoid(self, device, dtype):
+    @dtypes(torch.float, torch.double)
+    def test_hardsigmoid(self, device, dtype):
+        inputValues = [-1000, -4, -3, -2, 0, 2, 3, 4, 1000]
+        expectedOutput = [0.0, 0.0, 0.0, 0.1667, 0.5, 0.8333, 1.0, 1.0, 1.0]
```
Can we implement a numpy reference formula here instead of hard-coding results?

jamesr66a reviewed (commented):
Nice! I added a few comments inline. I'd also be interested to see the results from the microbenchmark, afaict we should be able to get this to peak machine memory bandwidth

vkuzo committed:
Update on "add hardsigmoid FP operator to PyTorch"


Summary:

This is for common operator coverage, since this is widely used.  A future PR
will add the quantized version.

Some initial questions for reviewers, since it's my first FP operator
diff:
* do we need a backwards.out method for this?
* do we need CUDA? If yes, should it be this PR or is it ok to split

Test Plan:

```
// test
python test/test_torch.py TestTorchDeviceTypeCPU.test_hardsigmoid_cpu_float32

// benchmark
python -m pt.hardsigmoid_test
...
Forward Execution Time (us) : 40.315

Forward Execution Time (us) : 42.603
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20371692](https://our.internmc.facebook.com/intern/diff/D20371692)

[ghstack-poisoned]

vkuzo commented:
> I'd also be interested to see the results from the microbenchmark, afaict we should be able to get this to peak machine memory bandwidth

from the benchmark I added, here is the full output: https://gist.github.com/vkuzo/4ab3067a3466a8155c6b07c57b346663.  Or, are you referencing the more comprehensive benchmarks you outlined in https://fb.quip.com/fZ9ZAcswrbYJ?  If yes, do we know if there is any reusable code already, or am I the first one

jamesr66a reviewed (approved):
@vkuzo I think the benchmark you added is fine, we can calculate the relevant memory bandwidth:

1) (3 * 256 * 256 * 4 bytes * 2) / 39.942e-6 seconds / 1e9 -> ~39 GiB/s
2) (4 * 3 * 256 * 256 * 4 bytes * 2) / 42.922e-6 seconds / 1e9 -> 147 GiB/s

That second figure indicates to me that there's threading happening. Do you mind rerunning the benchmark with environment variable `OMP_NUM_THREADS=1` set?

Otherwise, looks good to me!

vkuzo commented:
> @vkuzo I think the benchmark you added is fine, we can calculate the relevant memory bandwidth:
> 
> 1. (3 * 256 * 256 * 4 bytes * 2) / 39.942e-6 seconds / 1e9 -> ~39 GiB/s
> 2. (4 * 3 * 256 * 256 * 4 bytes * 2) / 42.922e-6 seconds / 1e9 -> 147 GiB/s
> 
> That second figure indicates to me that there's threading happening. Do you mind rerunning the benchmark with environment variable `OMP_NUM_THREADS=1` set?
> 
> Otherwise, looks good to me!

ah, got it, thanks for explaining!  Here is a paste of no-threading compared to threading: https://gist.github.com/vkuzo/565c2913fcb5c49b7ec9bd871a1974e9.  With `OMP_NUM_THREADS=1` the N=4 runtime is ~ 4x the N=1 runtime

vkuzo committed:
Update on "add hardsigmoid FP operator to PyTorch"


Summary:

This is for common operator coverage, since this is widely used.  A future PR
will add the quantized version.

Some initial questions for reviewers, since it's my first FP operator
diff:
* do we need a backwards.out method for this?
* do we need CUDA? If yes, should it be this PR or is it ok to split

Test Plan:

```
// test
python test/test_torch.py TestTorchDeviceTypeCPU.test_hardsigmoid_cpu_float32

// benchmark
python -m pt.hardsigmoid_test
...
Forward Execution Time (us) : 40.315

Forward Execution Time (us) : 42.603
(full output: https://gist.github.com/vkuzo/4ab3067a3466a8155c6b07c57b346663)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20371692](https://our.internmc.facebook.com/intern/diff/D20371692)

[ghstack-poisoned]

vkuzo committed:
Update on "add hardsigmoid FP operator to PyTorch"


Summary:

This is for common operator coverage, since this is widely used.  A future PR
will add the quantized version.

Some initial questions for reviewers, since it's my first FP operator
diff:
* do we need a backwards.out method for this?
* do we need CUDA? If yes, should it be this PR or is it ok to split

Test Plan:

```
// test
python test/test_torch.py TestTorchDeviceTypeCPU.test_hardsigmoid_cpu_float32

// benchmark
python -m pt.hardsigmoid_test
...
Forward Execution Time (us) : 40.315

Forward Execution Time (us) : 42.603
(full output: https://gist.github.com/vkuzo/4ab3067a3466a8155c6b07c57b346663)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20371692](https://our.internmc.facebook.com/intern/diff/D20371692)

[ghstack-poisoned]

jamesr66a commented:
@vkuzo Cool, with those numbers, batch-size 1 runs at ~11.5GiB/s and batch size 4 runs at ~12.15GiB/s, which seems reasonable. It's about half as fast as I think machine peak would be, but it's fine for this PR

vkuzo committed:
Update on "add hardsigmoid FP operator to PyTorch"


Summary:

This is for common operator coverage, since this is widely used.  A future PR
will add the quantized version.

Some initial questions for reviewers, since it's my first FP operator
diff:
* do we need a backwards.out method for this?
* do we need CUDA? If yes, should it be this PR or is it ok to split

Test Plan:

```
// test
python test/test_torch.py TestTorchDeviceTypeCPU.test_hardsigmoid_cpu_float32

// benchmark
python -m pt.hardsigmoid_test
...
Forward Execution Time (us) : 40.315

Forward Execution Time (us) : 42.603
(full output: https://gist.github.com/vkuzo/4ab3067a3466a8155c6b07c57b346663)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20371692](https://our.internmc.facebook.com/intern/diff/D20371692)

[ghstack-poisoned]

z-a-f commented on benchmarks/operator_benchmark/pt/hardsigmoid_test.py:
```diff
@@ -0,0 +1,66 @@
+from __future__ import absolute_import
```
BTW, for the benchmarks there is some extra work that you need to go through from the FB side

z-a-f reviewed (commented):
This implementation assumes cutoffs at `[-3, +3]`, and the slope of `1.0 / 6.0`. Is there a reason for these numbers? Can we parametrize the slope or is it a requirement? Also, TF implementation has different numerics: https://www.tensorflow.org/api_docs/python/tf/keras/activations/hard_sigmoid

z-a-f commented on torch/nn/modules/activation.py:
```diff
@@ -271,6 +271,29 @@ def forward(self, input):
+    r"""Applies the element-wise function:
+
+    .. math::
+        \text{Hardsigmoid}(x) = \frac{ReLU6(x + 3)}{6}
```
This math takes a little longer to understand. Instead, we can write it as:

```
.. math::
    \text{Hardsigmoid}(x) = \begin{cases}
        0 & \text{if~} x \le -3, \\
        1 & \text{if~} x \ge +3, \\
        x / 6 & \text{otherwise}
    \end{cases}
```

would render into

![image](https://user-images.githubusercontent.com/4216323/76648146-5868b180-651b-11ea-9cfc-607a24bbd06e.png)

vkuzo commented:
> This implementation assumes cutoffs at `[-3, +3]`, and the slope of `1.0 / 6.0`. Is there a reason for these numbers? Can we parametrize the slope or is it a requirement? Also, TF implementation has different numerics: https://www.tensorflow.org/api_docs/python/tf/keras/activations/hard_sigmoid

ah, I was going with the original definition from https://arxiv.org/pdf/1905.02244.pdf, which also has FB usage with the same parameters (and no FB usage I found with other params).  But, it would be easy to parametrize it from -alpha to alpha.  Let me ask you offline if we should do it here, if we do happy to update.

z-a-f commented:
> > This implementation assumes cutoffs at `[-3, +3]`, and the slope of `1.0 / 6.0`. Is there a reason for these numbers? Can we parametrize the slope or is it a requirement? Also, TF implementation has different numerics: https://www.tensorflow.org/api_docs/python/tf/keras/activations/hard_sigmoid
> 
> ah, I was going with the original definition from https://arxiv.org/pdf/1905.02244.pdf, which also has FB usage with the same parameters (and no FB usage I found with other params). But, it would be easy to parametrize it from -alpha to alpha. Let me ask you offline if we should do it here, if we do happy to update.

Awesome, let's add the reference to the paper in the comment.

vkuzo committed:
Update on "add hardsigmoid FP operator to PyTorch"


Summary:

This is for common operator coverage, since this is widely used.  A future PR
will add the quantized version.

Some initial questions for reviewers, since it's my first FP operator
diff:
* do we need a backwards.out method for this?
* do we need CUDA? If yes, should it be this PR or is it ok to split

Test Plan:

```
// test
python test/test_torch.py TestTorchDeviceTypeCPU.test_hardsigmoid_cpu_float32

// benchmark
python -m pt.hardsigmoid_test
...
Forward Execution Time (us) : 40.315

Forward Execution Time (us) : 42.603
(full output: https://gist.github.com/vkuzo/4ab3067a3466a8155c6b07c57b346663)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20371692](https://our.internmc.facebook.com/intern/diff/D20371692)

[ghstack-poisoned]

dizcza commented:
Has this functional been added in the 1.5 docs? I cannot find it.

vkuzo commented:
> Has this functional been added in the 1.5 docs? I cannot find it.

Thanks for the catch! This isn't in the 1.5 docs as it was experimental at branch cut time - targeting 1.6.  It's ready in the nightlies.
